video: allow placement of the video buffer pool in a ZEPHYR REGION (and adapt STM32N6 / STM32H7RS config)

avolmat-st opened the pull request:
The video buffer pool is placed in the default memory region. Due to that it wasn't possible to allow video buffer in other memory and multi-heap / shared-multi-heap was used.
Another alternative is simply to be able to indicate in which region the video buffer pool should be placed.
Similar approach is now available for LVGL buffers via dedicated config, making usage under the hood of the Z_GENERIC_SECTION mechanism.

Similar approach is proposed here, via addition two new KConfig for video subsystem VIDEO_BUFFER_POOL_ZEPHYR_REGION and VIDEO_BUFFER_POOL_ZEPHYR_REGION_NAME, to indicate where the video buffer pool should be placed.

This gives more flexibility on the whole soc memory allocation strategy, by simply partitioning the memory via DT, giving required caching properties as well.
Tricky part is that, depending on the memory, it might not be available yet from the very beginning, making it impossible to use the normal Z_HEAP_DEFINE_BY_SECT macro since this would lead to configuration the heap too early during the boot at a moment when the memory isn't yet available.
To overcome this, this manually configure the k_heap via the k_heap_init on the very first allocation request.

Two other commits are to move away from multi-heap / shared-multi-heap for LTDC / video buffer on STM32 platforms.
(if this is agreed, it would probably be necessary to either remove the SMH configuration of the MEMC XSPI/OSPI  or at least make it configurable ?)

Commit:
video: allow placement of video buffer pool in a Zephyr region

Addition of two options in order to select the Zephyr region
into which the video buffer pool should be placed.
CONFIG_VIDEO_BUFFER_POOL_ZEPHYR_REGION allows to indicate that the
video video pool should be placed in a specific ZEPHYR region which
name is CONFIG_VIDEO_BUFFER_POOL_ZEPHYR_REGION_NAME

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
boards: stm32n6570_dk: use ZEPHYR_REGION instead of SMH

Switch to usage of ZEPHYR_REGION for display and video
buffers instead of usage via the Shared-Multi-Heap API.
The ZEPHYR_REGION mechanism offer more flexibility than
the shared-multi-heap which is allocated on the whole
PSRAM.
If enabled, add allocation of the VENC internal buffers
from the PSRAM via the HEAP allocator.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
boards: stm32h7s78_dk: use ext-sdram instead of SMH for display

Switch to usage of ext-sdram property (ZEPHYR_REGION) for display
buffers instead of usage via the Shared-Multi-Heap API.
The ZEPHYR_REGION mechanism offer more flexibility than the
shared-multi-heap which is allocated on the whole PSRAM.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

josuah commented:
It definitely sounds like an improvement, thank you!

Currently we have `CONFIG_VIDEO_BUFFER_POOL_SZ_MAX * CONFIG_VIDEO_BUFFER_POOL_NUM_MAX` to set the size, but currently this does not work (allocation fails) because there is overhead for the heap header struct, for alignment, and fragmentation can happen on top so `malloc()/free()` can randomly make the allocation fail at runtime (which is helped by reusing buffers).

Maybe a different API would be `CONFIG_VIDEO_BUFFER_POOL_SIZE=` computed by the user.

I have the impression that what users would want is `video_buf_alloc(dev);`, then with `video_buf_alloc()`:

- If it asks the number of buffers via a CID
- It asks the alignment requirements via a CID
- It asks the buffer size via the video API
- It enqueues the video buffers right away
- It uses the memory region from Kconfig, or an extra parameter for an `_smh()` variant maybe.

What you proposed does not seem to require any API change, so it is probably not a problem to change the internals frequently... Just an opportunity to rework things.

ngphibang commented:
> Currently we have CONFIG_VIDEO_BUFFER_POOL_SZ_MAX * CONFIG_VIDEO_BUFFER_POOL_NUM_MAX to set the size, but currently this does not work (allocation fails) because there is overhead for the heap header struct, for alignment, and fragmentation can happen on top so malloc()/free() can randomly make the allocation fail at runtime (which is helped by reusing buffers).

I think this is because one of the two Kconfig exposed to user, the `CONFIG_VIDEO_BUFFER_POOL_SZ_MAX` does not make sense and cause the confusion, i.e., in fact, if we allocate less than `CONFIG_VIDEO_BUFFER_POOL_NUM_MAX`, the size of one buffer can still be greater than `CONFIG_VIDEO_BUFFER_POOL_SZ_MAX`. A simple fix is to expose

- `CONFIG_VIDEO_BUFFER_POOL_NUM_MAX` : max numbers of buffers (for video_buf[] array)
- `CONFIG_VIDEO_BUFFER_POOL_HEAP_SIZE` : size in bytes of the video buffer pool. This way user can specify the total size and does not need to adjust the header struct to their buffer size, etc.

Commit:
video: add CONFIG_VIDEO_BUFFER_POOL_HEAP_SIZE in configs

In preparation to the renaming of CONFIG_VIDEO_BUFFER_POOL_SZ_MAX
into CONFIG_VIDEO_BUFFER_POOL_HEAP_SIZE, add the new CONFIG
in all conf files equal to POOL_SZ_MAX multiply by POOL_NUM_MAX.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
video: add CONFIG_VIDEO_BUFFER_POOL_HEAP_SIZE description

Clarify the size of the video buffer pool by having a
dedicated CONFIG for it.  Until now the size of the
video buffer pool was equal to VIDEO_BUFFER_POOL_SZ_MAX
multiply by VIDEO_BUFFER_POOL_NUM_MAX.

This commit only add the description, the config doesn't
have yet any effect. Change will be added after all configs
are updated to define it in order to avoid breaking
platforms between 2 commits.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
video: replace CONFIG_VIDEO_BUFFER_POOL_SZ_MAX with POOL_HEAP_SIZE

Update video common code and applications to rely on the
CONFIG_VIDEO_BUFFER_POOL_HEAP_SIZE instead of
CONFIG_VIDEO_BUFFER_POOL_SZ_MAX.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
video: remove CONFIG_VIDEO_BUFFER_POOL_SZ_MAX in conf files

Remove all CONFIG_VIDEO_BUFFER_POOL_SZ_MAX settings in
config files.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
video: removal of VIDEO_BUFFER_POOL_SZ_MAX from Kconfig

Now that VIDEO_BUFFER_POOL_HEAP_SIZE is available is used
in all projects, VIDEO_BUFFER_POOL_SZ_MAX can be removed.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
samples: video: tcpserversink: fix stm32 venc max buffer size

The CONFIG_VIDEO_BUFFER_POOL_SZ_MAX represent the size of the
biggest buffer in the pool size and not the whole pool size.
For the stm32-venc this should be 1-000-000 and not 10-000-000.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
video: stm32-venc: allow selection of venc internal memory pool

Allow usage of either Shared-Multi-Heap based internal memory pool
allocation or allocation from a HEAP located optional in a
Zephyr region.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
video: fix VIDEO_BUFFER_USE_SHARED_MULTI_HEAP definition

Avoid default n since this is the default for a bool config

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

etienne-lms commented on drivers/video/video_stm32_venc.c:
```diff
@@ -16,8 +16,14 @@
+#define EWL_HEAP_ALIGNED_ALLOC(size)\
+	shared_multi_heap_aligned_alloc(CONFIG_VIDEO_BUFFER_SMH_ATTRIBUTE, ALIGNMENT_INCR, (size))
+#define EWL_HEAP_ALIGNED_FREE(block) shared_multi_heap_free(block)
+#endif
```
Is mixing header files inclusion and macro definition really nice? Always including **shared_multi_heap.h** should not be an issue. Suggestion
```patch
 #include <zephyr/logging/log.h>
 #include <zephyr/multi_heap/shared_multi_heap.h>

 #include <ewl.h>
 
 (...)

 #define ALIGNMENT_INCR 8UL

+#if defined(CONFIG_VIDEO_STM32_VENC_SMH_ALLOC)
 #define EWL_HEAP_ALIGNED_ALLOC(size)\
 	shared_multi_heap_aligned_alloc(CONFIG_VIDEO_BUFFER_SMH_ATTRIBUTE, ALIGNMENT_INCR, (size))
 #define EWL_HEAP_ALIGNED_FREE(block) shared_multi_heap_free(block)
+#endif /* CONFIG_VIDEO_STM32_VENC_SMH_ALLOC */

ngphibang commented on samples/subsys/usb/uvc/prj.conf:
```diff
@@ -11,6 +11,6 @@ CONFIG_USB_DEVICE_STACK_NEXT=y
 CONFIG_CDC_ACM_SERIAL_INITIALIZE_AT_BOOT=n
 CONFIG_VIDEO=y
 CONFIG_VIDEO_BUFFER_POOL_NUM_MAX=2
```
ditto

ngphibang commented on tests/drivers/video/test_pattern/src/main.c:
```diff
@@ -206,7 +206,7 @@ static void *test_pattern_setup(void)
 	zassert(caps.min_vbuf_count <= CONFIG_VIDEO_BUFFER_POOL_NUM_MAX,
 		"not enough buffers");
-	zassert(fmt.size <= CONFIG_VIDEO_BUFFER_POOL_SZ_MAX,
+	zassert(fmt.size <= (CONFIG_VIDEO_BUFFER_POOL_HEAP_SIZE / CONFIG_VIDEO_BUFFER_POOL_NUM_MAX),
```
ditto

josuah commented on doc/releases/release-notes-4.4.rst:
```diff
@@ -173,6 +156,12 @@ New APIs and options
+
+  * :kconfig:option:`CONFIG_VIDEO_BUFFER_POOL_HEAP_SIZE`
+  * :kconfig:option:`CONFIG_VIDEO_BUFFER_POOL_ZEPHYR_REGION`
+  * :kconfig:option:`CONFIG_VIDEO_BUFFER_POOL_ZEPHYR_REGION_NAME`
```
It is tempting to shorten them as:
- `CONFIG_VIDEO_BUFFER_POOL_SIZE` (or `_BUFFER_HEAP_SIZE`?)
- `CONFIG_VIDEO_BUFFER_REGION` (or `_BUFFER_MEMORY_REGION`?)
- `CONFIG_VIDEO_BUFFER_REGION_NAME`

ngphibang commented:
> Should I keep those 5 commits related to POOL_HEAP_SIZE or squash all into 1 commit (which would make conf update / framework update at once which I do not feel right), or 2 commits (one for framework update and 1 for conf updates) but in this case this will break during 1 commit so not good from a bisect point of view.

Yes, we need to keep bisectability. Keeping 5 commits or squashing into 1 commit are both fine to me (personally, I prefer one commit because it is simpler).

Just still a typo in the commit title  "video: allow placement of video buffer **pool** in a Zephyr region"

Commit:
video: remove CONFIG_VIDEO_BUFFER_POOL_NUM_MAX=2 from conf files

VIDEO_BUFFER_POOL_NUM_MAX default value is 2 so remove it from
samples conf files setting it to that default value.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

avolmat-st commented on samples/subsys/usb/uvc/src/main.c:
```diff
@@ -100,7 +100,7 @@ static int app_add_format(uint32_t pixfmt, uint32_t width, uint32_t height, bool
 	}
 
-	if (fmt.size > CONFIG_VIDEO_BUFFER_POOL_SZ_MAX) {
+	if (fmt.size > (CONFIG_VIDEO_BUFFER_POOL_HEAP_SIZE / CONFIG_VIDEO_BUFFER_POOL_NUM_MAX)) {
```
@ngphibang, @josuah, actually, looking more into detail how it is currently done within the UVC sample, allocation is done as following:

within the main function:
```c
        uint32_t uvc_buf_count = CONFIG_VIDEO_BUFFER_POOL_NUM_MAX;
```
then, based on this that number is either fully allocated for the camera side, or splitted into 2 and half allocated for the camera and half for the encoder.

So, based on that, I am not sure we can make something more generic than what is currently done / proposed which is:

```c
if (fmt.size > (CONFIG_VIDEO_BUFFER_POOL_HEAP_SIZE / CONFIG_VIDEO_BUFFER_POOL_NUM_MAX))
```

Of course that one doesn't take into consideration if there is an encoder or not and the size generated by the encoder, but this might still be a good enough approximation no ?

avolmat-st commented on samples/drivers/video/capture/boards/arduino_nicla_vision_stm32h747xx_m7.conf:
```diff
@@ -1,5 +1,5 @@
+CONFIG_VIDEO_BUFFER_POOL_HEAP_SIZE=320000
 CONFIG_VIDEO_BUFFER_POOL_NUM_MAX=2
```
Done

erwango reviewed (dismissed):
Approved from STM32 side.
A rebase is required, though

JarmouniA reviewed (commented):
Nit: I believe you should combine the following commits into 1:
- video: add CONFIG_VIDEO_BUFFER_POOL_HEAP_SIZE in configs
- video: replace CONFIG_VIDEO_BUFFER_POOL_SZ_MAX with POOL_HEAP_SIZE
- video: remove CONFIG_VIDEO_BUFFER_POOL_SZ_MAX in conf files
